app/src/main/java/org/wikipedia/richtext/CustomHtmlParser.kt

```diff
@@ -92,8 +108,55 @@ class CustomHtmlParser constructor(private val handler: TagHandler) : TagHandler
         private var lastAClass = ""
 
         override fun handleTag(opening: Boolean, tag: String?, output: Editable?, attributes: Attributes?): Boolean {
-            if (tag == "img") {
+            if (tag == "img" && view == null) {
                 return true
+            } else if (tag == "img" && opening && view != null) {
+                var imgWidth = DimenUtil.htmlPxToInt(getValue(attributes, "width").orEmpty())
+                var imgHeight = DimenUtil.htmlPxToInt(getValue(attributes, "height").orEmpty())
+                val imgSrc = getValue(attributes, "src").orEmpty()
+
+                if (imgWidth < MIN_IMAGE_SIZE || imgHeight < MIN_IMAGE_SIZE) {
+                    return true
+                }
+
+                imgWidth = DimenUtil.roundedDpToPx(imgWidth.toFloat())
+                imgHeight = DimenUtil.roundedDpToPx(imgHeight.toFloat())
+
+                if (imgWidth > 0 && imgHeight > 0 && imgSrc.isNotEmpty()) {
+                    val bmpMap = contextBmpMap.getOrPut(view.context) { mutableMapOf() }
+                    var drawable = bmpMap[imgSrc]
+
+                    if (drawable == null || drawable.bitmap.isRecycled) {
+                        // give it a placeholder drawable of the appropriate size
+                        drawable = BitmapDrawable(view.context.resources,
+                            Bitmap.createBitmap(imgWidth, imgHeight, Bitmap.Config.RGB_565))
+                        bmpMap[imgSrc] = drawable
+
+                        drawable.setBounds(0, 0, imgWidth, imgHeight)
+
+                        var uri = imgSrc
+                        if (uri.startsWith("//")) {
+                            uri = WikiSite.DEFAULT_SCHEME + ":" + uri
+                        } else if (uri.startsWith("./")) {
+                            uri = Service.COMMONS_URL + uri.replace("./", "")
+                        }
+
+                        Glide.with(view)
+                            .asBitmap()
+                            .load(uri)
+                            .into(object : CustomTarget<Bitmap>() {
+                                override fun onResourceReady(resource: Bitmap, transition: Transition<in Bitmap>?) {
+                                    if (!drawable.bitmap.isRecycled) {
+                                        drawable.bitmap.applyCanvas {
+                                            drawBitmap(resource, Rect(0, 0, resource.width, resource.height), drawable.bounds, null)
```

Comment:
Can we add `WhiteBackgroundTransformation().maybeDimImage(drawable.bitmap)` here to respect the image dimming setting in the app?

Reply:
yes!